File: go/encoding.go
package turbotoken

import (
	"sync"
)

// Encoding holds a loaded BPE encoding with its rank data, ready for
// encoding, decoding, and counting operations.
type Encoding struct {
	name        string
	spec        *EncodingSpec
	rankPayload []byte
	once        sync.Once
	loadErr     error
}

// newEncoding creates an Encoding from a spec, deferring rank file loading.
func newEncoding(spec *EncodingSpec) *Encoding {
	return &Encoding{
		name: spec.Name,
		spec: spec,
	}
}

// ensureLoaded loads the rank file if not already loaded.
func (enc *Encoding) ensureLoaded() error {
	enc.once.Do(func() {
		enc.rankPayload, enc.loadErr = ReadRankFile(enc.name)
	})
	return enc.loadErr
}

// Name returns the encoding name.
func (enc *Encoding) Name() string {
	return enc.name
}

// NVocab returns the vocabulary size.
func (enc *Encoding) NVocab() int {
	return enc.spec.NVocab
}

// Encode tokenizes text into BPE token IDs.
func (enc *Encoding) Encode(text string) ([]uint32, error) {
	if err := enc.ensureLoaded(); err != nil {
		return nil, err
	}
	return ffiEncodeBPE(enc.rankPayload, []byte(text))
}

// Decode converts BPE token IDs back to a UTF-8 string.
func (enc *Encoding) Decode(tokens []uint32) (string, error) {
	if err := enc.ensureLoaded(); err != nil {
		return "", err
	}
	data, err := ffiDecodeBPE(enc.rankPayload, tokens)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// Count returns the number of BPE tokens in text without materializing the token array.
func (enc *Encoding) Count(text string) (int, error) {
	if err := enc.ensureLoaded(); err != nil {
		return 0, err
	}
	return ffiCountBPE(enc.rankPayload, []byte(text))
}

// CountTokens is an alias for Count.
func (enc *Encoding) CountTokens(text string) (int, error) {
	return enc.Count(text)
}

// IsWithinTokenLimit checks if text tokenizes to at most limit tokens.
// Returns (tokenCount, true, nil) if within limit, (0, false, nil) if exceeded.
func (enc *Encoding) IsWithinTokenLimit(text string, limit int) (int, bool, error) {
	if err := enc.ensureLoaded(); err != nil {
		return 0, false, err
	}
	return ffiIsWithinTokenLimit(enc.rankPayload, []byte(text), limit)
}

// EncodeChat encodes a sequence of chat messages into BPE tokens using
// the configured chat template.
func (enc *Encoding) EncodeChat(messages []ChatMessage, opts *ChatOptions) ([]uint32, error) {
	formatted := formatChat(messages, opts)
	return enc.Encode(formatted)
}

// CountChat counts BPE tokens for a chat message sequence.
func (enc *Encoding) CountChat(messages []ChatMessage, opts *ChatOptions) (int, error) {
	formatted := formatChat(messages, opts)
	return enc.Count(formatted)
}

// IsChatWithinTokenLimit checks if chat messages tokenize within a limit.
func (enc *Encoding) IsChatWithinTokenLimit(messages []ChatMessage, limit int, opts *ChatOptions) (int, bool, error) {
	formatted := formatChat(messages, opts)
	return enc.IsWithinTokenLimit(formatted, limit)
}

// EncodeFilePath encodes the contents of a file into BPE tokens.
func (enc *Encoding) EncodeFilePath(path string) ([]uint32, error) {
	if err := enc.ensureLoaded(); err != nil {
		return nil, err
	}
	return ffiEncodeBPEFile(enc.rankPayload, path)
}

// CountFilePath counts BPE tokens in a file without materializing the token array.
func (enc *Encoding) CountFilePath(path string) (int, error) {
	if err := enc.ensureLoaded(); err != nil {
		return 0, err
	}
	return ffiCountBPEFile(enc.rankPayload, path)
}

File: csharp/Encoding.cs
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Text;

namespace TurboToken
{
    /// <summary>
    /// A BPE encoding instance backed by a rank file.
    /// </summary>
    public sealed class Encoding : IDisposable
    {
        /// <summary>The encoding name.</summary>
        public string Name { get; }

        /// <summary>The encoding spec.</summary>
        public EncodingSpec Spec { get; }

        private byte[] _rankPayload;
        private GCHandle _rankHandle;
        private bool _disposed;

        internal Encoding(string name, EncodingSpec spec, byte[] rankPayload)
        {
            Name = name;
            Spec = spec;
            _rankPayload = rankPayload;
            _rankHandle = GCHandle.Alloc(_rankPayload, GCHandleType.Pinned);
        }

        private IntPtr RankPtr => _rankHandle.AddrOfPinnedObject();
        private UIntPtr RankLen => (UIntPtr)_rankPayload.Length;

        // MARK: - Encode

        /// <summary>Encode text to BPE token IDs.</summary>
        public uint[] Encode(string text)
        {
            ThrowIfDisposed();
            var textBytes = Encoding8.GetBytes(text);
            return CallTwoPassUInt32(textBytes, (txtPtr, txtLen, outPtr, outCap) =>
                NativeMethods.turbotoken_encode_bpe_from_ranks(
                    RankPtr, RankLen, txtPtr, txtLen, outPtr, outCap));
        }

        /// <summary>Decode BPE token IDs back to a UTF-8 string.</summary>
        public string Decode(uint[] tokens)
        {
            ThrowIfDisposed();
            var bytes = CallTwoPassUInt8(tokens, (tokPtr, tokLen, outPtr, outCap) =>
                NativeMethods.turbotoken_decode_bpe_from_ranks(
                    RankPtr, RankLen, tokPtr, tokLen, outPtr, outCap));
            return Encoding8.GetString(bytes);
        }

        // MARK: - Count

        /// <summary>Count the number of BPE tokens for the given text.</summary>
        public int Count(string text)
        {
            ThrowIfDisposed();
            var textBytes = Encoding8.GetBytes(text);
            unsafe
            {
                fixed (byte* txtPtr = textBytes)
                {
                    var result = NativeMethods.turbotoken_count_bpe_from_ranks(
                        RankPtr, RankLen,
                        (IntPtr)txtPtr, (UIntPtr)textBytes.Length);
                    var count = result.ToInt64();
                    if (count < 0)
                        throw new TurboTokenException($"count returned error code {count}");
                    return (int)count;
                }
            }
        }

        /// <summary>Alias for Count.</summary>
        public int CountTokens(string text) => Count(text);

        // MARK: - Token Limit

        /// <summary>
        /// Check if text is within a token limit.
        /// Returns the token count if within limit.
        /// Throws TokenLimitExceededException if exceeded.
        /// </summary>
        public int IsWithinTokenLimit(string text, int limit)
        {
            ThrowIfDisposed();
            var textBytes = Encoding8.GetBytes(text);
            unsafe
            {
                fixed (byte* txtPtr = textBytes)
                {
                    var result = NativeMethods.turbotoken_is_within_token_limit_bpe_from_ranks(
                        RankPtr, RankLen,
                        (IntPtr)txtPtr, (UIntPtr)textBytes.Length,
                        (UIntPtr)limit);
                    var code = result.ToInt64();
                    if (code == -2)
                        throw new TokenLimitExceededException(limit);
                    if (code < 0)
                        throw new TurboTokenException($"isWithinTokenLimit returned error code {code}");
                    return (int)code;
                }
            }
        }

        // MARK: - Chat

        /// <summary>Encode a chat conversation to token IDs.</summary>
        public uint[] EncodeChat(IReadOnlyList<ChatMessage> messages, ChatOptions? options = null)
        {
            var text = ChatFormatter.FormatChat(messages, options ?? new ChatOptions());
            return Encode(text);
        }

        /// <summary>Count tokens in a chat conversation.</summary>
        public int CountChat(IReadOnlyList<ChatMessage> messages, ChatOptions? options = null)
        {
            var text = ChatFormatter.FormatChat(messages, options ?? new ChatOptions());
            return Count(text);
        }

        /// <summary>Check if a chat conversation is within a token limit.</summary>
        public int IsChatWithinTokenLimit(IReadOnlyList<ChatMessage> messages, int limit, ChatOptions? options = null)
        {
            var text = ChatFormatter.FormatChat(messages, options ?? new ChatOptions());
            return IsWithinTokenLimit(text, limit);
        }

        // MARK: - File Operations

        /// <summary>Encode a file's contents to BPE token IDs.</summary>
        public uint[] EncodeFilePath(string path)
        {
            ThrowIfDisposed();
            var pathBytes = Encoding8.GetBytes(path);
            return CallTwoPassUInt32(pathBytes, (pPtr, pLen, outPtr, outCap) =>
                NativeMethods.turbotoken_encode_bpe_file_from_ranks(
                    RankPtr, RankLen, pPtr, pLen, outPtr, outCap));
        }

        /// <summary>Count tokens in a file.</summary>
        public int CountFilePath(string path)
        {
            ThrowIfDisposed();
            var pathBytes = Encoding8.GetBytes(path);
            unsafe
            {
                fixed (byte* pPtr = pathBytes)
                {
                    var result = NativeMethods.turbotoken_count_bpe_file_from_ranks(
                        RankPtr, RankLen,
                        (IntPtr)pPtr, (UIntPtr)pathBytes.Length);
                    var count = result.ToInt64();
                    if (count < 0)
                        throw new TurboTokenException($"countFilePath returned error code {count}");
                    return (int)count;
                }
            }
        }

        /// <summary>Check if a file's content is within a token limit.</summary>
        public int IsFilePathWithinTokenLimit(string path, int limit)
        {
            ThrowIfDisposed();
            var pathBytes = Encoding8.GetBytes(path);
            unsafe
            {
                fixed (byte* pPtr = pathBytes)
                {
                    var result = NativeMethods.turbotoken_is_within_token_limit_bpe_file_from_ranks(
                        RankPtr, RankLen,
                        (IntPtr)pPtr, (UIntPtr)pathBytes.Length,
                        (UIntPtr)limit);
                    var code = result.ToInt64();
                    if (code == -2)
                        throw new TokenLimitExceededException(limit);
                    if (code < 0)
                        throw new TurboTokenException($"isFilePathWithinTokenLimit returned error code {code}");
                    return (int)code;
                }
            }
        }

        // MARK: - Two-pass helpers

        private delegate IntPtr TwoPassUInt32Fn(IntPtr input, UIntPtr inputLen, IntPtr outBuf, UIntPtr outCap);
        private delegate IntPtr TwoPassUInt8Fn(IntPtr input, UIntPtr inputLen, IntPtr outBuf, UIntPtr outCap);

        private unsafe uint[] CallTwoPassUInt32(byte[] input, TwoPassUInt32Fn fn)
        {
            fixed (byte* inPtr = input)
            {
                var sizeResult = fn((IntPtr)inPtr, (UIntPtr)input.Length, IntPtr.Zero, UIntPtr.Zero);
                var count = sizeResult.ToInt64();
                if (count < 0)
                    throw new TurboTokenException($"FFI size query returned error code {count}");
                if (count == 0)
                    return Array.Empty<uint>();

                var buffer = new uint[count];
                fixed (uint* outPtr = buffer)
                {
                    var written = fn((IntPtr)inPtr, (UIntPtr)input.Length, (IntPtr)outPtr, (UIntPtr)count);
                    var writtenCount = written.ToInt64();
                    if (writtenCount < 0)
                        throw new TurboTokenException($"FFI fill returned error code {writtenCount}");
                    if (writtenCount < count)
                        Array.Resize(ref buffer, (int)writtenCount);
                }
                return buffer;
            }
        }

        private unsafe byte[] CallTwoPassUInt8(uint[] input, TwoPassUInt8Fn fn)
        {
            fixed (uint* inPtr = input)
            {
                var sizeResult = fn((IntPtr)inPtr, (UIntPtr)input.Length, IntPtr.Zero, UIntPtr.Zero);
                var count = sizeResult.ToInt64();
                if (count < 0)
                    throw new TurboTokenException($"FFI size query returned error code {count}");
                if (count == 0)
                    return Array.Empty<byte>();

                var buffer = new byte[count];
                fixed (byte* outPtr = buffer)
                {
                    var written = fn((IntPtr)inPtr, (UIntPtr)input.Length, (IntPtr)outPtr, (UIntPtr)count);
                    var writtenCount = written.ToInt64();
                    if (writtenCount < 0)
                        throw new TurboTokenException($"FFI fill returned error code {writtenCount}");
                    if (writtenCount < count)
                        Array.Resize(ref buffer, (int)writtenCount);
                }
                return buffer;
            }
        }

        private static readonly System.Text.Encoding Encoding8 = System.Text.Encoding.UTF8;

        // MARK: - IDisposable

        private void ThrowIfDisposed()
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(Encoding));
        }

        public void Dispose()
        {
            if (!_disposed)
            {
                _disposed = true;
                if (_rankHandle.IsAllocated)
                    _rankHandle.Free();
            }
        }
    }
}
